tests/integration/ha/test_ha.py

```diff
@@ -197,6 +197,77 @@ async def test_kill_db_process_node_with_primary_shard(
     await assert_continuous_writes_consistency(c_writes)
 
 
+@pytest.mark.abort_on_fail
+async def test_freeze_db_process_node_with_primary_shard(
+    ops_test: OpsTest, c_writes: ContinuousWrites, c_balanced_writes_runner
+) -> None:
+    """Check cluster can self-heal + data indexed/read on process freeze on node with P_shard."""
+    app = (await app_name(ops_test)) or APP_NAME
+
+    units_ips = get_application_unit_ids_ips(ops_test, app)
+    leader_unit_ip = await get_leader_unit_ip(ops_test, app=app)
+
+    # find unit hosting the primary shard of the index "series-index"
+    shards = await get_shards_by_index(ops_test, leader_unit_ip, ContinuousWrites.INDEX_NAME)
+    first_unit_with_primary_shard = [shard.unit_id for shard in shards if shard.is_prim][0]
+
+    # Killing the only instance can be disastrous.
+    if len(ops_test.model.applications[app].units) < 2:
```

Comment:
Seems like you do this throughout the code I think it would be cleaner to add this as a function - or even better as a fixture

Reply:
Thanks! I got some feedback in a previous PR to make them obvious in the test as opposed to hiding them in a function, and I agree with it as it should ideally only there be 1 place where unit scaling happens (easier to read / debug)

Reply:
Makes sense! Thanks for sharing this :)